### DISCUSSION TITLE
Named deconstruction

### DISCUSSION THREAD

**Discussion by AlexRadch** [dotnet/csharplang#1895] [Language Ideas]
Deconstruction uses positions syntax only and has syntax for skipping certain positions.
I suggest to add a named deconstruction syntax.

Named deconstruction enable you to specify correct property for a variable by associating the variable name with the property's name rather than with the property's position in deconstruction.
Also it allow to skip some properties in deconstruction.
```
            var points = new (int x, int y, int z)[] { (0, 0, 0), (1, 1, 1) };
            foreach ((var px: x, var py: y) in points) { }
            foreach ((var py: y, var pz: z) in points) { }
            foreach ((var pz: z, var py: y, var px: x) in points) { }
```

**Comment by hez2010** [comment]
Actually you can write it already:
```c#
foreach (var (px, py, pz) in points) {   }
```

**Comment by AlexRadch** [comment]
@hez2010 but you can not write `foreach (var (pz: z, py: y, px: x) in points) {   }`. Also you should remember each position. Also code will become wrong if properties order is changed.
It is like https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/classes-and-structs/named-and-optional-arguments

**Comment by Austin-bryan** [comment]
I always love when the existing syntax is shorter than purposed syntax. And you should really try a better example, because if someone is order `x`, `y`, and `z`, in, frankly, literally anyother order, then you have a different problem altogether.
